flask/launcher/src/main/java/net/corda/flask/launcher/JarCache.java

```diff
@@ -0,0 +1,226 @@
+                return false;
+            } else {
+                log.debug("Using cache directory '{}'", candidate.toString());
+                return true;
+            }
+        } catch (Exception ioe) {
+            log.debug(
+                    String.format("Cache directory '%s' discarded: %s", candidate.toString(), ioe.getMessage()),
+                    ioe
+            );
+            return false;
+        }
+    }
+
+    @SneakyThrows
+    private static Path computeCacheDirectory(String appName) {
+        Stream<Optional<Path>> candidates;
+        if (OS.isUnix) {
+            candidates = Stream.of(
+                    Optional.ofNullable(System.getProperty(Flask.JvmProperties.CACHE_DIR)).map(Paths::get),
+                    Optional.ofNullable(System.getenv("XDG_CACHE_HOME")).map(prefix -> Paths.get(prefix, appName)),
+                    Optional.ofNullable(System.getProperty("user.home")).map(prefix -> Paths.get(prefix, ".cache", appName)),
+                    Optional.ofNullable(System.getProperty("java.io.tmpdir")).map(path -> Paths.get(path).resolve(appName)),
+                    Optional.of(Paths.get("/tmp", appName))
+            );
+        } else if (OS.isMac) {
+            candidates = Stream.of(
+                    Optional.ofNullable(System.getProperty(Flask.JvmProperties.CACHE_DIR)).map(Paths::get),
+                    Optional.ofNullable(System.getProperty("user.home"))
+                            .map(prefix -> Paths.get(prefix, "Library", "Saved Application State", appName)),
+                    Optional.of(Paths.get("/Library/Caches", appName)),
+                    Optional.ofNullable(System.getProperty("java.io.tmpdir"))
+                            .map(prefix -> Paths.get(prefix).resolve(appName)));
+        } else if (OS.isWindows) {
+            candidates = Stream.of(
+                    Optional.ofNullable(System.getProperty(Flask.JvmProperties.CACHE_DIR)).map(Paths::get),
+                    Optional.ofNullable(System.getenv())
+                            .map(it -> it.get("LOCALAPPDATA"))
+                            .map(prefix -> Paths.get(prefix, appName)),
+                    Optional.ofNullable(System.getProperty("user.home"))
+                            .map(prefix -> Paths.get(prefix, "AppData", "Local", appName)),
+                    Optional.ofNullable(System.getProperty("user.home"))
+                            .map(prefix -> Paths.get(prefix, "Local Settings", "Application Data", appName)),
+                    Optional.ofNullable(System.getProperty("java.io.tmpdir"))
+                            .map(prefix -> Paths.get(prefix).resolve(appName))
+            );
+
+        } else {
+            candidates = Stream.of(
+                    Optional.ofNullable(System.getProperty(Flask.JvmProperties.CACHE_DIR)).map(Paths::get),
+                    Optional.ofNullable(System.getProperty("java.io.tmpdir")).map(path -> Paths.get(path).resolve(appName))
+            );
```

Comment:
Aren't these two options common to all cases? And then you can use [Stream.concat()](https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html#concat-java.util.stream.Stream-java.util.stream.Stream-).

Reply:
I'll get rid of this and just use the same set of candidates directories on all OS